WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -1090,20 +1090,28 @@ private async Task<IEnumerable<TxOut>> ProceedWithOutputRegistrationPhaseAsync(u
 		// Get the output's size and its of the input that will spend it in the future.
 		// Here we assume all the outputs share the same scriptpubkey type.
 		var isTaprootAllowed = roundParameters.AllowedOutputTypes.Contains(ScriptType.Taproot);
-		var preferTaprootOutputs = isTaprootAllowed && Random.Shared.NextDouble() < .5;
-		var (inputVirtualSize, outputVirtualSize) = DestinationProvider.Peek(preferTaprootOutputs).IsScriptType(ScriptType.Taproot)
-			? (Constants.P2trInputVirtualSize, Constants.P2trOutputVirtualSize)
-			: (Constants.P2wpkhInputVirtualSize, Constants.P2wpkhOutputVirtualSize);
 
-		AmountDecomposer amountDecomposer = new(roundParameters.MiningFeeRate, roundParameters.AllowedOutputAmounts, outputVirtualSize, inputVirtualSize, (int)availableVsize);
+		AmountDecomposer amountDecomposer = new(roundParameters.MiningFeeRate, roundParameters.AllowedOutputAmounts, (int)availableVsize, isTaprootAllowed);
 		var theirCoins = constructionState.Inputs.Where(x => !registeredCoins.Any(y => x.Outpoint == y.Outpoint));
 		var registeredCoinEffectiveValues = registeredAliceClients.Select(x => x.EffectiveValue);
 		var theirCoinEffectiveValues = theirCoins.Select(x => x.EffectiveValue(roundParameters.MiningFeeRate, roundParameters.CoordinationFeeRate));
 		var outputValues = amountDecomposer.Decompose(registeredCoinEffectiveValues, theirCoinEffectiveValues);
 
 		// Get as many destinations as outputs we need.
-		var destinations = DestinationProvider.GetNextDestinations(outputValues.Count(), preferTaprootOutputs).ToArray();
-		var outputTxOuts = outputValues.Zip(destinations, (amount, destination) => new TxOut(amount, destination.ScriptPubKey));
+		List<TxOut> outputTxOuts = new();
+		foreach (var output in outputValues)
+		{
+			var needTaprootOutput = output.ScriptType is ScriptType.Taproot;
+			var dest = DestinationProvider.GetNextDestinations(1, needTaprootOutput).First();
+
+			var txOut = new TxOut(output.Amount, dest.ScriptPubKey);
+			if (!txOut.ScriptPubKey.IsScriptType(needTaprootOutput ? ScriptType.Taproot : ScriptType.P2WPKH))
+			{
+				throw new InvalidOperationException("ScriptType mismatch.");
+			}
```

Comment:
I believe this is not necessary. I had a confusion and then confused others with the "preferTaproot" thing but that was fixed.

Reply:
Removed